fuse_constraints/include/fuse_constraints/normal_delta_pose_3d_cost_functor.h

```diff
@@ -0,0 +1,189 @@
+ *
+ * where, mu is a vector and S is a covariance matrix, then, A = S^{-1/2}, i.e the matrix A is the square root
+ * information matrix (the inverse of the covariance).
+ *
+ * Note that the cost function's quaternion components are only concerned with the imaginary components (qx, qy, qz).
+ */
+class NormalDeltaPose3DCostFunctor
+{
+public:
+  /**
+   * @brief Constructor
+   *
+   * @param[in] A The residual weighting matrix, most likely the square root information matrix in order
+   *              (dx, dy, dz, dqx, dqy, dqz)
+   * @param[in] b The exposed pose difference in order (dx, dy, dz, dqw, dqx, dqy, dqz)
+   */
+  NormalDeltaPose3DCostFunctor(const fuse_core::Matrix6d& A, const fuse_core::Vector7d& b);
+
+  /**
+   * @brief Compute the cost values/residuals using the provided variable/parameter values
+   */
+  template <typename T>
+  bool operator()(
+    const T* const position1,
+    const T* const orientation1,
+    const T* const position2,
+    const T* const orientation2,
+    T* residual) const;
+
+private:
+  fuse_core::Matrix6d A_;  //!< The residual weighting matrix, most likely the square root information matrix
+  fuse_core::Vector7d b_;  //!< The measured difference between variable pose1 and variable pose2
+};
+
+NormalDeltaPose3DCostFunctor::NormalDeltaPose3DCostFunctor(const fuse_core::Matrix6d& A, const fuse_core::Vector7d& b) :
+  A_(A),
+  b_(b)
+{
+}
+
+template <typename T>
+bool NormalDeltaPose3DCostFunctor::operator()(
+  const T* const position1,
+  const T* const orientation1,
+  const T* const position2,
+  const T* const orientation2,
+  T* residual) const
+{
+  // Based on the Ceres SLAM pose graph error calculation here:
+  // https://github.com/ceres-solver/ceres-solver/blob/4fc5d25f9cbfe2aa333425ddad03bdc651335c24/examples/slam/pose_graph_3d/pose_graph_3d_error_term.h#L73
+
+  // 1. Get the relative orientation change from the variable's orientation1 to orientation2
+  T variable_orientation1_inverse[4] =
+  {
+    orientation1[0],
+    -orientation1[1],
+    -orientation1[2],
+    -orientation1[3]
+  };
+
+  T variable_orientation_delta[4];
+  ceres::QuaternionProduct(variable_orientation1_inverse, orientation2, variable_orientation_delta);
+
+  // 2. Get the position change from pose1 to pose2, then rotate it into the frame of pose1
+  T variable_position_delta[3] =
+  {
+    position2[0] - position1[0],
+    position2[1] - position1[1],
+    position2[2] - position1[2]
+  };
+
+  T variable_position_delta_rotated[3];
+  ceres::QuaternionRotatePoint(
+    variable_orientation1_inverse,
+    variable_position_delta,
+    variable_position_delta_rotated);
+
+  // 3. Get the difference between the orientation delta that we just computed for orientation1 and orientation2,
+  // and the measurement's orientation delta
+  T observation_orientation_delta[4] =
+  {
+    T(b_(3)),
+    T(b_(4)),
+    T(b_(5)),
+    T(b_(6))
+  };
+
+  T variable_orientation_delta_inverse[4] =
```

Comment:
Hrm, strangely, no. This is an odd one, because they are also following the standard of `error = variable - measurement`, and I believe I am doing the same thing here that they were.

https://github.com/ceres-solver/ceres-solver/blob/master/examples/slam/pose_graph_3d/pose_graph_3d_error_term.h#L103
https://github.com/ceres-solver/ceres-solver/blob/master/examples/slam/pose_graph_3d/pose_graph_3d_error_term.h#L96

Bizarrely, putting them the way we did for the orientation variable has zero effect on the outcome; the covariance matrix is utterly identical. Given that one form is the inverse of the other, I find this very strange.

Reply:
Could they have this wrong in theirs? Check this Octave script out. I can provide the functions, if you want to see those:

```
% Standard is error = variable - measurement

quat_variable = rpy_to_quat(0, 0, 0.2);
quat_measurement = rpy_to_quat(0, 0, 0.1);

quat_delta1 = quat_mult(quat_variable, quat_conj(quat_measurement));  % What we do for the orientation constraint
quat_delta2 = quat_mult(quat_measurement, quat_conj(quat_variable));  % The way this code is now

[r1 p1 y1] = quat_to_rpy(quat_delta1)    % This one should be correct
[r2 p2 y2] = quat_to_rpy(quat_delta2)
```

Sure enough:

```
r1 = 0
p1 = 0
y1 = 0.1

r2 = 0
p2 = 0
y2 = -0.1
```

So I think we need to change this here. Since it apparently doesn't affect the results in my test, I feel both good and bad about that.